### ISSUE TITLE
Using nbclient to talk to jupyter lab  running remotely

### ISSUE THREAD

**Issue by amit-chandak-unskript** [jupyter/nbclient#213]
Hi,
      I have a use case wherein i have jupyterlab server running on an EC2 instance and i want to run a .ipynb file against a kernel inside that jupyterlab. I was wondering if i can use nbclient to achieve that? I have used nbclient to talk to enterprise gateway and [run notebooks](https://github.com/jupyter-server/enterprise_gateway/issues/972) but when i try the same approach for standalone jupyterlab server, it doesnt work.

**Comment by davidbrochart** [comment]
Hi @amit-chandak-unskript,
I think you will need to [pass a kernel manager to nbclient](https://github.com/jupyter/nbclient/blob/86dcb64e18de6cec3050ca7c2a36ad7519000958/nbclient/client.py#L413-L415), with a custom kernel provisioner that allows to talk to the kernel remotely. If you don't want to control the life cycle of this kernel, but only execute code, I think you just need to get the kernel's connection info. The ZMQ sockets are TCP sockets that can be accessed remotely, so this should work. But I don't know if this has been done before, maybe @kevin-bates has ideas?

**Comment by kevin-bates** [comment]
Thanks @davidbrochart, hello @amit-chandak-unskript.

Hmm. One of the primary differences between running a kernel via the jupyter-server (lab) REST API and the gateway REST API, is that jupyter-server is _session-centric_, while the gateway's are more _kernel-centric_, meaning that all things start from a session in jupyter-server whereas the gateway doesn't.  What happens when you use the `GatewayKernelManager` with `nbclient` and point at your jupyter-server instance?   

Might you be able to deploy (and expose) a gateway instance beside your lab instance in EC2?  They could both share the same kernel specifications, but have their own managed space (i.e., Lab won't see the Gateway's kernels and vice versa).

**Comment by amit-chandak-unskript** [comment]
Its not reachability issue,  as i confirmed the jupyterlab server url works if i do the following
```
 curl -XGET https://<jupyterlab base url>/api/kernelspecs
{"default": "python3", "kernelspecs": {"python3": {"name": "python3", "spec": {"argv": ["/opt/conda/bin/python", "-m", "ipykernel_launcher", "-f", "{connection_file}"], "env": {}, "display_name": "python3", "language": "python", "interrupt_mode": "signal", "metadata": {"debugger": true}}, "resources": {"logo-32x32": "/d3c71c1b-d075-4b5c-a6dd-9f417f483c3e/kernelspecs/python3/logo-32x32.png", "logo-64x64": "/d3c71c1b-d075-4b5c-a6dd-9f417f483c3e/kernelspecs/python3/logo-64x64.png"}}}}(con
```

**Comment by amit-chandak-unskript** [comment]
@kevin-bates i like your idea of having an enterprise gateway beside the jupyterlab server with same kernelspec and use that for nbclient. But i would like to avoid maintaining 2 instances. So, nbclient was never meant to be used with remote jupyterlab server, is it?

**Comment by davidbrochart** [comment]
> So, nbclient was never meant to be used with remote jupyterlab server, is it?

No, it wasn't. Nbclient doesn't talk any HTTP, and jupyter-server is an HTTP server. The only web that it talks is TCP sockets, which are used to connect to a (remote) kernel. If it had access to the kernel's connection info, that would be possible.

**Comment by amit-chandak-unskript** [comment]
Thanks @davidbrochart , one more question, is it possible to make nbclient use a custom kernel to run? as in does it take a kernel spec to run the notebook?

**Comment by davidbrochart** [comment]
Yes, you will need to pass your own kernel manager ([km](https://github.com/jupyter/nbclient/blob/20081a6b0ad28d2a3fcc8fc508e8ccf9e08d510c/nbclient/client.py#L406)). You can get one like this:
```python
from jupyter_client.manager import KernelManager

km = KernelManager(kernel_name="python3")
km.start_kernel()

# pass km to nbclient

km.shutdown_kernel()
```

**Comment by kevin-bates** [comment]
Right, but, as [David points out](https://github.com/jupyter/nbclient/issues/213#issuecomment-1068488433), this won't get you to the remote server.  You'd essentially be rewriting `GatewayKernelManager` (and `GatewayKernelClient`).  Since you received a 403, you might need to explore adding the token to your headers via `JUPYTER_GATEWAY_AUTH_TOKEN`.  I suspect there will be other issues as well - probably in the session management aspect of things.

**Comment by davidbrochart** [comment]
@amit-chandak-unskript I'm working on a new project that will allow to do just what you want, see https://github.com/davidbrochart/jpterm/pull/2.
